### PR TITLE
GIS: random bumps

### PR DESCRIPTION
**Pull request by dotlambda**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

The PROJ bump breaks many packages. Maybe I should only make it use CMake for now and wait a little longer until I update it?

cc @mpickering @johannesloetzsch @tollb

### CODE REVIEW

**Review by sikmir — APPROVED** [review]

**Comment by das-g** [comment]
Can we keep proj 7 available in parallel until upstream projects have caught up in supporting proj 8?

**Comment by das-g** [comment]
Should fa01607858c4d09ccf549181d7884c349fe73942 be reordered to after 2b7686be9577b7f70158a6bb54ae1c09ec5e4258, so that all intermediate revisions are also buildable?

**Comment by dotlambda** [comment]
> Should [fa01607](https://github.com/NixOS/nixpkgs/commit/fa01607858c4d09ccf549181d7884c349fe73942) be reordered to after [2b7686b](https://github.com/NixOS/nixpkgs/commit/2b7686be9577b7f70158a6bb54ae1c09ec5e4258), so that all intermediate revisions are also buildable?

I guess that's a good idea in case we ever need `git bisect`.

**Review by lsix — CHANGES_REQUESTED** [review]
Hi,

Thanks for doing that! Last time I updated `proj`, I left out the 8.0.0 update because of the removal of `proj_api.h` which creates some breakage.

I am still rebuilding packages with your changes, but at the moment I can see that at least `spatialite_tools` seems to break because of it.

Here is a fix for the problem if you apply the patch on top of your branch (I did not check the formatting):

```patch
diff --git a/pkgs/development/libraries/spatialite-tools/default.nix b/pkgs/development/libraries/spatialite-tools/default.nix
index ffded04510b..69c897377c8 100644
--- a/pkgs/development/libraries/spatialite-tools/default.nix
+++ b/pkgs/development/libraries/spatialite-tools/default.nix
@@ -1,15 +1,16 @@
-{ lib, stdenv, fetchurl, pkg-config, sqlite, expat, zlib, proj, geos, libspatialite, readosm }:
+{ lib, stdenv, fetchurl, pkg-config, sqlite, expat, zlib, proj, geos, libspatialite, readosm, libxml2, minizip, librttopo }:
 
 stdenv.mkDerivation rec {
-  name = "spatialite-tools-4.1.1";
+  pname = "spatialite-tools";
+  version = "5.0.0";
 
   src = fetchurl {
-    url = "https://www.gaia-gis.it/gaia-sins/spatialite-tools-sources/${name}.tar.gz";
-    sha256 = "14aqmhvab63ydbb82fglsbig7jw1wmci8jjvci07aavdhvh1pyrv";
+    url = "https://www.gaia-gis.it/gaia-sins/spatialite-tools-sources/${pname}-${version}.tar.gz";
+    sha256 = "0ckddgdpxhy6vkpr9q2hnx5qmanrd8g4pqnifbrq1i5jrj82s2dd";
   };
 
   nativeBuildInputs = [ pkg-config ];
-  buildInputs = [ sqlite expat zlib proj geos libspatialite readosm ];
+  buildInputs = [ sqlite expat zlib proj geos libspatialite readosm libxml2 minizip librttopo ];
 
   configureFlags = [ "--disable-freexl" ];
 
```

Otherwise, from what I can see, it looks good to me. I’ll try to finish the rebuilds to see if I can spot other breakages.

**Comment by lsix** [comment]
`python*Packages.cartopy` is also broken by this update and would need `proj7` I guess, but I am not sure how viable a cohabitation with libraries that dependent on `proj-8.x.x` will be within the same python interpreter, I’ll have to check.

File: pkgs/applications/science/misc/gplates/default.nix
Comment (dotlambda) [inline pkgs/applications/science/misc/gplates/default.nix:15]: This means GPlates is using PROJ 7, but also using GDAL which in turn uses PROJ 8. Should we use `gdal.override { proj = proj7 }` instead?

**Comment by dotlambda** [comment]
> `python*Packages.cartopy` is also broken by this update and would need `proj7` I guess, but I am not sure how viable a cohabitation with libraries that dependent on `proj-8.x.x` will be within the same python interpreter, I’ll have to check.

It shouldn't create a Python-related problem, but I don't know if some internal data format is incompatible or whatever.

File: pkgs/servers/sql/postgresql/ext/postgis.nix
Comment (dotlambda) [inline pkgs/servers/sql/postgresql/ext/postgis.nix:27]: Same problem as for GPlates.

File: pkgs/applications/science/misc/gplates/default.nix
Comment (kmicklas) [reply pkgs/applications/science/misc/gplates/default.nix:15]: That does seem like the safer option, if GPlates doesn't work with 8.

Comment (dotlambda) [reply pkgs/applications/science/misc/gplates/default.nix:15]: Actually, I don't think that's a good idea: It creates unnecessary load on Hydra and actually we'd need to override GDAL's dependencies as well: GDAL depends on libspatialite, which depend on PROJ. That's a rabbit hole I don't wanna get into.

**Comment by dotlambda** [comment]
@lsix I'm having trouble updating QGIS to the latest version which should have support for PROJ 8. Would you mind having a look? Feel free to push to my branch.

**Comment by lsix** [comment]
@dotlambda I’ll have a look asap

**Comment by dotlambda** [comment]
Result of `nixpkgs-review pr 124299` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages marked as broken and skipped:</summary>
  <ul>
    <li>googleearth-pro</li>
    <li>grass</li>
    <li>openspace</li>
    <li>python38Packages.geopandas</li>
    <li>python38Packages.osmnx</li>
    <li>python39Packages.basemap</li>
    <li>python39Packages.geopandas</li>
    <li>python39Packages.osmnx</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>qgis</li>
    <li>qgis-unwrapped</li>
  </ul>
</details>
<details>
  <summary>52 packages built:</summary>
  <ul>
    <li>apacheHttpdPackages.mod_tile (apacheHttpdPackages_2_4.mod_tile)</li>
    <li>cloudcompare</li>
    <li>gdal (python38Packages.gdal)</li>
    <li>gdal_2</li>
    <li>gmt</li>
    <li>gplates</li>
    <li>libLAS</li>
    <li>libgeotiff</li>
    <li>librttopo</li>
    <li>libspatialite</li>
    <li>mapnik</li>
    <li>mapproxy</li>
    <li>merkaartor</li>
    <li>mysql-workbench</li>
    <li>octavePackages.octproj</li>
    <li>openorienteering-mapper</li>
    <li>osmscout-server</li>
    <li>paraview</li>
    <li>pdal</li>
    <li>perl530Packages.Tirex</li>
    <li>perl532Packages.Tirex</li>
    <li>postgresqlPackages.postgis (postgresql11Packages.postgis)</li>
    <li>proj</li>
    <li>proj7</li>
    <li>python38Packages.basemap</li>
    <li>python38Packages.cartopy</li>
    <li>python38Packages.fiona</li>
    <li>python38Packages.owslib</li>
    <li>python38Packages.pygmt</li>
    <li>python38Packages.pyproj</li>
    <li>python38Packages.python-mapnik</li>
    <li>python38Packages.rasterio</li>
    <li>worldengine-cli (python38Packages.worldengine)</li>
    <li>python39Packages.cartopy</li>
    <li>python39Packages.fiona</li>
    <li>python39Packages.gdal</li>
    <li>python39Packages.owslib</li>
    <li>python39Packages.pygmt</li>
    <li>python39Packages.pyproj</li>
    <li>python39Packages.python-mapnik</li>
    <li>python39Packages.rasterio</li>
    <li>python39Packages.worldengine</li>
    <li>pytrainer</li>
    <li>qlandkartegt</li>
    <li>qmapshack</li>
    <li>readosm</li>
    <li>saga</li>
    <li>spatialite_tools</li>
    <li>sumo</li>
    <li>udig</li>
    <li>valhalla</li>
    <li>xastir</li>
  </ul>
</details>

**Comment by dotlambda** [comment]
@lsix There is version 3.18.3.

**Comment by lsix** [comment]
@dotlambda I know, but I try to stay to the LTR version in nixpkgs as much as possible, which currently is 3.16.7. I guess we could have a `qgis-latest` or something latest, but otherwise it sometimes is more difficult to backport things to stable release branches (at least it was when I started opting for the LTR releases).

I also have added the nixpkgs bump in your branch, but since it is not strictly required for the `proj-8` update, it could also go in a dedicated branch, depending on your preference.

**Comment by lsix** [comment]
@dotlambda I have also added a `postgis` update to your branch. It fixes the build with the updated `proj` and avoid relying on `proj7`.

**Comment by tollb** [comment]
@dotlambda I noticed an issue with python-mapnik. The python interpreter core dumps on exit if you "import mapnik". I believe this is due to linking with both proj 7.2.1 and proj 8.0.1. Try:

nix-shell -p python3Packages.python-mapnik --run 'python3 -c "import mapnik"'

I think there are at least two conflicts:

1. The gdal package which is required by mapnik is using proj 8.0.1, while mapnik still requires proj7.
2. The python-mapnik package references the proj package (which would be proj 8.0.1). I think that reference may be unnecessary -- or could be updated to use proj7.

**Comment by dotlambda** [comment]
@tollb Do you think it's an option to switch to an unreleased version of mapnik or ask upstream to make a new release?

File: pkgs/applications/gis/qgis/unwrapped.nix
Comment (jonringer) [inline pkgs/applications/gis/qgis/unwrapped.nix:40]: do you mind commenting as why this is necessary?

File: pkgs/development/libraries/libspatialite/default.nix
Comment (jonringer) [inline pkgs/development/libraries/libspatialite/default.nix:29]: probably doesn't matter since it's set to `null` on linux, but libiconv should be a darwin only

File: pkgs/top-level/all-packages.nix
Comment (FRidh) [inline pkgs/top-level/all-packages.nix:17447]: proj_7

**Review by FRidh — CHANGES_REQUESTED** [review]

**Comment by tollb** [comment]
> @tollb Do you think it's an option to switch to an unreleased version of mapnik or ask upstream to make a new release?

@dotlambda Good questions. I don't have complete answers, but here is some background:
1. The last mapnik release (3.1.0) was in January, and there has only been one release a year the last few years.
2. It appears that there will be additional work required in python-mapnik to support the new proj api, which is planned by upstream. See: https://github.com/mapnik/mapnik/pull/4202#issuecomment-808166105. There is a proj6 branch for python-mapnik, but the work hasn't been merged to master: https://github.com/mapnik/python-mapnik/commits/proj6
3. It looks like Debian sid is still building mapnik 3.1.0 with proj 7.2.1.
4. It appears that arch linux is trying to build mapnik 3.1.0 with proj 8, but I'm not sure that it successfully includes proj support.

Although there are upstream patches, it may not be a good idea to get ahead of upstream (or other major distributions) since we probably lack the ability to test these packages effectively; the new proj support has required a large set of changes for mapnik.

Are there other options?

**Comment by das-g** [comment]
> Are there other options?

Would it (for now) be easier to keep `proj` as proj 7 and have `proj_8` for packages requiring the newer proj 8? (I guess the latter group is currently limited to only the bumped `qmapshack`, or are there others?)

**Comment by dotlambda** [comment]
> > Are there other options?
> 
> Would it (for now) be easier to keep `proj` as proj 7 and have `proj_8` for packages requiring the newer proj 8? (I guess the latter group is currently limited to only the bumped `qmapshack`, or are there others?)

Why would we do that? We shouldn't enslave ourselves to poorly maintained upstream projects.
If I understand it correctly, the API change was done in PROJ 6, the old API was deprecated in PROJ 7 and removed in PROJ 8.
I see no reason not to mark these packages as broken and leave it up to their users to replace `proj` with `proj_7` in an overlay.

**Comment by dotlambda** [comment]
I'm giving up on the PROJ update for now.

**Comment by das-g** [comment]
> > > Are there other options?
> > 
> > Would it (for now) be easier to keep `proj` as proj 7 and have `proj_8` for packages requiring the newer proj 8? [...]
> 
> Why would we do that?

I'm not saying we _should_ do that, I'm only suggesting it as an option, because there was a question about "other options". The upside _might_ be that we can keep more packages not-broken with less work / headaches. Off course, that'd only be the case if it leads to fewer version conflicts for packages that depend on proj via several ways. Whether or not that'd be the case I don't know; it's what I wanted to ask by "Would it be easier [...]?".

> We shouldn't enslave ourselves to poorly maintained upstream projects.

Of course not. If we went that way, it'd have to remain a temporary measure until a sufficient number of upstream projects have caught up. Then `proj` would become proj 8 and a `proj_7` would only be kept around if absolutely needed.

> If I understand it correctly, the API change was done in PROJ 6, the old API was deprecated in PROJ 7 and removed in PROJ 8.

It seems the old API and the old header `<proj_api.h>` was deprecated even already with 5.0.0, though it's not clear to me whether the API provided by the new header `<proj.h>`changed again significantly from 5.x to 6. Indeed the old API and header were removed with proj 8.

But the new API didn't just mean different function names and function signatures: The new API reflects a new model how the library works and should be used (see ["Background" in the guide for "Version 4 to 5 API Migration"](https://proj.org/development/migration.html#background)), so depending on how the library was used, adopting the new API could have been non-trivial and might have been postponed even by otherwise relatively well-maintained, if though maybe not very fast-paced upstream projects. (See e.g. https://github.com/osmcode/libosmium/issues/295#issuecomment-615875721, though libosmium wants to eventually get rid of proj altogether, if I understood correctly.)

(I'm not here to defend these upstream projects or their upstream maintainers, just to offer a plausible explanation why so many projects depending on proj seem to still depend on the old API.)

> I see no reason not to mark these packages as broken and leave it up to their users to replace `proj` with `proj_7` in an overlay.

If there's a reason to bump `proj` already now, we should definitely do that. Though AFAIK the old proj versions don't have any known security issues and more nixpkgs users might use proj through one of the projects still depending on the old API than directly, so postponing the bump and not forcing the work of configuring an overlay on those users might be more user friendly.

But as already written: If that approach doesn't lead to fewer version conflicts, let's just bump `proj` anyway.

**Comment by dotlambda** [comment]
Btw, the prior work can still be found at https://github.com/dotlambda/nixpkgs/tree/proj-8.0.1.

**Comment by tollb** [comment]
The issue of mixing proj7 and gdal (built with proj 8) encountered with mapnik _might_ affect some other packages at runtime. Searching nixpkgs as it exists in the PR:

$ git grep -l -e proj7 --and -e gdal
pkgs/applications/misc/merkaartor/default.nix
pkgs/applications/misc/qlandkartegt/default.nix
pkgs/applications/science/misc/gplates/default.nix
pkgs/servers/sql/postgresql/ext/postgis.nix

Maybe we could consider introducing an alternate version of gdal built with proj7? I don't know enough about the pros and cons of this approach to make a recommendation.

FYI, Debian has setup a tracker for their proj transition showing some dependency information. It may be of interest, especially as they make updates.
https://release.debian.org/transitions/html/auto-proj.html

**Comment by lsix** [comment]
> I'm giving up on the PROJ update for now.

This is also where I got last time I updated PROJ :cry: However, the amount of incompatible packages has decreased since, so things are going in the right direction!

Thanks @dotlambda  for the work you have put in this, it is definitely not lost, and thanks for the other feedbacks as well!

I’ll relaunch builds tomorrow and re-read everything, but this PR seems in a good state as it is.

**Comment by dotlambda** [comment]
QMapShack fails with
```
-- Found Threads: TRUE
-- Found GDAL: /nix/store/fjry6cwbjh69d7wqyybkzldznx31vzjs-gdal-3.2.2/lib/libgdal.so (found version "3.2.2")
-- Found JPEG: /nix/store/0byg8crnnyg1wki90c9r9ffcv423q5f4-libjpeg-turbo-2.0.6/lib/libjpeg.so (found version "62")
-- Found ROUTINO: /nix/store/yfm1npjxfwkf60fv4dwhhhikfwsljzy8-routino-3.3.3/lib/libroutino.so
-- Found ROUTINO: /nix/store/yfm1npjxfwkf60fv4dwhhhikfwsljzy8-routino-3.3.3/include
-- Found ROUTINO: /nix/store/yfm1npjxfwkf60fv4dwhhhikfwsljzy8-routino-3.3.3/share/routino
-- Found ZLIB: /nix/store/65ys3k6gn2s27apky0a0la7wryg3az9q-zlib-1.2.11/lib/libz.so (found version "1.2.11")
-- Could not find ALGLIB using internal code.
CMake Error at CMakeLists.txt:166 (message):
  You need at least PROJ 7.2.1 or newer.
```
which cofuses me a lot.

**Comment by dotlambda** [comment]
I fixed QMapShack. It was actually a problem with PROJ's CMake stuff: https://github.com/OSGeo/PROJ/issues/2731.

It seems like there are some issues with PROJ's tests on Darwin. I guess it has to do with the Darwin sandbox. Can someone with macOS system look into that?

**Comment by dotlambda** [comment]
@ofborg build proj

**Comment by dotlambda** [comment]
Alright, I seem to have solved the problem on Darwin. I'll build all depending packages and then this should be good to go.

File: pkgs/applications/gis/qmapshack/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/gis/qmapshack/default.nix:27]: Please delete this file.

File: pkgs/development/libraries/librttopo/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/librttopo/default.nix:15]: Does gitea work with fetchFromGitHub? Maybe we should use fetchGit instead.

File: pkgs/development/libraries/proj/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/proj/default.nix:50]:
```suggestion
    platforms = platforms.unix;
```

File: pkgs/development/libraries/librttopo/default.nix
Comment (dotlambda) [reply pkgs/development/libraries/librttopo/default.nix:15]: Yes, it's compatible.
I could add
```
fetchFromGitea = { domain, ... }@args: fetchFromGitHub (args // { githubBase = domain; })
```

**Comment by dotlambda** [comment]
@lsix Please merge if this looks good to you.

**Review by lsix — APPROVED** [review]
LGTM, built everything.

Thanks for the work!

File: pkgs/development/libraries/proj/default.nix
Comment (sikmir) [inline pkgs/development/libraries/proj/default.nix:30]: @dotlambda pkg-config file (proj.pc) not installed when using CMake (fixed in 8.0.0, https://github.com/OSGeo/PROJ/pull/2547). That's problem, since proj is used by many projects (and out of nixpkgs) and many of them use pkg-config, not cmake.

**Pull request merged**
